### PR TITLE
[experiment] Report incorrect cancellations

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#pragma warning disable CA1068 // CancellationToken parameters must come last
+        internal static CancellationTokenSource CreateLinkedTokenSource(CancellationToken cancellationToken1, CancellationToken cancellationToken2, [CallerFilePath] string? sourcePath = null, [CallerLineNumber] int sourceLine = 0)
+            => RecordLocation(CancellationTokenSource.CreateLinkedTokenSource(cancellationToken1, cancellationToken2), sourcePath, sourceLine);
+#pragma warning restore
+
+        internal static bool TryGetCreationSourceLocation(CancellationToken cancellationToken, [NotNullWhen(true)] out string? sourcePath, out int sourceLine)
+        {
+            var source = (CancellationTokenSource?)typeof(CancellationToken)
+                .GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
+                .FirstOrDefault(field => field.FieldType == typeof(CancellationTokenSource))?
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs:45]: extract fieldinfo lookup to a static?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        internal static string GetUnexpectedCancellationMessage(OperationCanceledException exception, CancellationToken cancellationToken)
+        {
+            var sourceLocation = TryGetCreationSourceLocation(exception.CancellationToken, out var sourcePath, out var sourceLine) ?
+                $"{sourcePath}({sourceLine})" : "unknown";
+
+            return $"Unexpected cancellation triggered by cancellation source at {sourceLocation}: " +
+                (cancellationToken == CancellationToken.None ? " caller does not expect cancellation" : "caller expects different token signaled");
+        }
+    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs:67]: cool idea :)

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by sharwell** [comment]
Marked as draft since I believe the intent here was an experiment. It looks like we missed a source:

```
    Microsoft.CodeAnalysis.Editor.UnitTests.FindReferences.FindReferencesTests.TestCSharp_Indexer_CRef(kind: API, host: OutOfProcess) [STARTING]
      System.InvalidOperationException : Unexpected cancellation triggered by cancellation source at unknown:  caller does not expect cancellation
      ---- System.Threading.Tasks.TaskCanceledException : A task was canceled.
      Stack Trace:
        /_/src/Workspaces/Remote/Core/BrokeredServiceConnection.cs(243,0): at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.OnUnexpectedException(Exception exception, CancellationToken cancellationToken)
        /_/src/Workspaces/Remote/Core/BrokeredServiceConnection.cs(106,0): at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__9.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        /_/src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs(97,0): at Microsoft.CodeAnalysis.Remote.RemoteHostClient.<TryInvokeAsync>d__13`1.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
        /_/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_FindReferences_Current.cs(43,0): at Microsoft.CodeAnalysis.FindSymbols.SymbolFinder.<FindReferencesAsync>d__34.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        /_/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_FindReferences_Legacy.cs(94,0): at Microsoft.CodeAnalysis.FindSymbols.SymbolFinder.<FindReferencesAsync>d__40.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
        /_/src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb(273,0): at Microsoft.CodeAnalysis.Editor.UnitTests.FindReferences.FindReferencesTests.VB$StateMachine_654_TestAPI.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        /_/src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb(43,0): at Microsoft.CodeAnalysis.Editor.UnitTests.FindReferences.FindReferencesTests.VB$StateMachine_646_TestAPIAndFeature.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        /_/src/EditorFeatures/Test2/FindReferences/FindReferencesTests.IndexerSymbols.vb(230,0): at Microsoft.CodeAnalysis.Editor.UnitTests.FindReferences.FindReferencesTests.VB$StateMachine_211_TestCSharp_Indexer_CRef.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        ----- Inner Stack Trace -----
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__143.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__133`1.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        /_/src/Workspaces/Remote/Core/BrokeredServiceConnection.cs(101,0): at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__9.MoveNext()
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            Contract.ThrowIfNull(sourcePath);
+            s_tokenCreators.Add(source, Tuple.Create(sourcePath, sourceLine));
+            return source;
+        }
+
+        public static CancellationTokenSource Create([CallerFilePath] string? sourcePath = null, [CallerLineNumber] int sourceLine = 0)
+            => RecordLocation(new CancellationTokenSource(), sourcePath, sourceLine);
+
+#pragma warning disable CA1068 // CancellationToken parameters must come last
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs:31]: 📝 This was recently fixed with https://github.com/dotnet/roslyn-analyzers/pull/4228

**Comment by tmat** [comment]
> It looks like we missed a source.

It might also be a source that we do not own - like JSON-RPC.

**Pull request closed without merging**